IFrame API

gjn opened the pull request:
This is a reduced alternative for https://github.com/geoadmin/mf-geoadmin3/issues/3146 and replaces https://github.com/geoadmin/mf-geoadmin3/pull/3164. It also establishes a first version of an iframe API.

It adds a service to handle iframe messaging to parent pages. Message is an object of the form
```
{
  type: messageType, //the message type
  payload: data //the data for this type
}
```

The PR also adds the notooltip parameter. If true, no tooltip is ever shown. It's application wide setting and applied for all layers. It's also working on mobile, desktop and embeded.

Existing message is adapted to make it structured. It adds the following messages:

- Selection of kml object
- Saving of kml file (as in https://github.com/geoadmin/mf-geoadmin3/pull/3317)
- Application change state (permalink is updated)

For details, see [Testlink (on codepen)](http://codepen.io/gjn19/pen/VajQyo)

gjn commented:
@davidoesch @cedricmoullet Please verify. This is for wandern.ch.

cedricmoullet commented:
1. I had to update the codepen (id's of the span: http://codepen.io/cedricmoullet/pen/BKzPwY Is it correctly changed ?
2. Why do you work with the src code //mf-geoadmin3.int.bgdi.ch/gjn_kml_external/src/embed.html  ?

gjn commented:
That's certainly not the one I wanted to link to.

cedricmoullet commented:
I tried to create a drawing with map.geo.admin.ch. It's possible to click on a text, but not on polygon and symbol. Any idea why ? (see http://codepen.io/cedricmoullet/pen/BKzPwY )

cedricmoullet commented:
I got it -> because of tooltip

gjn commented:
Codepen is now correct.

cedricmoullet commented:
Also if a tooltip is available, nothing is shown for symbol and polyline/polygon: http://codepen.io/cedricmoullet/pen/BKzPwY Any idea why ?

gjn commented:
Yes, seems like our kmls (created by drawing) don't contain name for the kml and no name for Placement, as described in the PR description.

On the other hand, we read those values for other kmls for the popup.

Looks like we need to treat differernt kml's differently. That's bad.

davidoesch commented:
"name for the kml" 
-> shall we use by default the hash generated by geo.admin.ch
 "name for Placement"
-> here we have to add names while drawing (Line1, line2 like google mymaps does st
<Placemark>        <name>Point 1</name>
instead of 
 <Placemark id="linepolygon_1457714969707

gjn commented:
We don't put anything in <name> tag for Placemarks because AFAIK, the name tag is autoamatically displayed when displaying it in google maps.

I think quick brainstorm is needed to define what we want to support. Remember, this is `hidden` feature (we don't document it)

davidoesch commented:
Brainstorm on Wed. As soon as wandern. Ch uses it... It will be hard to hide it

oterral commented:
Are you sure you don't want to send all the feature in a GeoJSON format ? That fit all usecases.
(@loicgasser is right send an object doesn't work on iE9)

gjn commented:
yes. KISS principle. layer id and feature id should be enough.

davidoesch commented:
is this already on prod? since it is already "documented" on http://api3.geo.admin.ch/releasenotes/index.html

gjn commented:
It's nowhere on official documentation. Codepn exmaple above is under my private account, not official codepen.

davidoesch commented:
![unbenannt](https://cloud.githubusercontent.com/assets/4577727/13858392/c421ecf8-ec7e-11e5-8f94-7b4da8b9efa2.PNG)
@gjn

gjn commented:
ouch. correct example (geojson layer, not kml layer), wrong text in api3 
page.

On 17.03.2016 20:28, David Oesch wrote:
> [1]
>  @gjn [2]
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub [3]
> 
> 	*
> 
> Links:
> ------
> [1]
> https://cloud.githubusercontent.com/assets/4577727/13858392/c421ecf8-ec7e-11e5-8f94-7b4da8b9efa2.PNG
> [2] https://github.com/gjn
> [3] 
> https://github.com/geoadmin/mf-geoadmin3/pull/3199#issuecomment-198046646

gjn commented:
@davidoesch We still need to decide if we want 

a) adapt what we send to parent window (use optional 'id' attribute instead of <name> tag)
or
b) adapt what we put in our own drawings

Because currently, our own drawings would not send anything when clicking an object because our placements have no <name> tag (because name tags are always displayed on google maps).

gjn commented:
I've adapted the PR. Now, when the Placemark has no name, we will transmit the 'id' attribute of the Placemark.

I've also added the notooltip permalink parameter. See PR description.

The codepen example has been updated to reflect these changes.

Anyway, we still have a related problem with drawings that we are creating with map.geo.admin.ch. If we load such a kml (red elements in the codepen example), most elements are not selectable, only the label element is. This is true outside of this PR as well. @oterral Do you know why this is the case? Could this easily be adapted?

For me, this PR is good to merge.

gjn commented:
ping @cedricmoullet @davidoesch Would be great to have this for next weeks deploy.

cedricmoullet commented:
1. I don't see in the code where the choice between "name" and "id" is made. Where it is ?
2. Any answer to the question you asked to @oterral ?

oterral commented:
Sorry guys. If a feature has no name and no description, it is not selectable because there is nothing to display in the tooltip.

gjn commented on src/components/tooltip/TooltipDirective.js:
```diff
@@ -490,6 +490,12 @@ goog.require('ga_topic_service');
               var name = feature.get('name');
               var featureId = feature.getId();
```
@cedricmoullet Per default, it's the id for a feature.

gjn commented on src/components/tooltip/TooltipDirective.js:
```diff
@@ -490,6 +490,12 @@ goog.require('ga_topic_service');
               var layerId = feature.get('layerId') || layer.get('bodId');
+              if (layer.get('type') == 'KML') {
+                layerId = layer.label;
+                if (name && name.length) {
```
@cedricmoullet For KML, and if the feature has a 'name' attribute, it will be the name.

gjn commented:
@oterral With the new 'notooltip' parameter, it would make sense to make those no-name/no-description objects selectable still.
Or in general, we make all kml objects selectable, but only display a tooltip if there is a name or description.

Thoughts?

oterral commented:
yes why not

gjn commented:
Ok. I've update the PR so that kml objects with not name/description are selectable - without a tooltip.

I've updated the codepen example accordingly.

@oterral Quick code review?

@davidoesch @cedricmoullet Good to merge?

davidoesch commented:
Testtes the codepen: ok from my side

oterral commented on src/components/tooltip/TooltipDirective.js:
```diff
@@ -490,12 +491,21 @@ goog.require('ga_topic_service');
                   replace('{{descr}}', feature.get('description') || '').
                   replace('{{name}}', (name) ? '(' + name + ')' : '');
               feature.set('htmlpopup', htmlpopup);
+              if (!isFeatureQueryable(feature)) {
```
If you want the same behavior in 3d you remove the test here : https://github.com/geoadmin/mf-geoadmin3/blob/369cb88b9c9aa3e341d2f6af1bbbce2318c6be50/src/components/tooltip/TooltipDirective.js#L347

oterral commented:
lgtm minor comment for 3d behavior

cedricmoullet commented:
Tested on mac (safari, ff, chrome), all fine.

davidoesch commented:
tested on Android  6.0.1  with chrome: ok

procrastinatio commented:
You are really opening the Pandora's box with iframe - parent's windows communication

gjn commented:
I've updated this PR with a couple of API possibilities and added structure to the messages passed.

For details, see [Testlink (on codepen)](http://codepen.io/gjn19/pen/VajQyo)

davidoesch commented:
Nice!-- lets see the pandoras box discussion

gjn commented:
Note: this will fix https://github.com/geoadmin/mf-geoadmin3/issues/3152

davidoesch commented:
As decided with @gjn , we make a silent release of this PR with the new structure. But we will remove the a) drawing save message and b) permalink update messages.

cedricmoullet commented:
To be sure: it doesn't affect hydrodaten, right ?

gjn commented:
Yes, for 98% of browsers.

We are not sure for IE9...I could not verify for IE9 because codepen doesn't work with IE9.

gjn commented:
I looked deeper in their code (which is written with the necessary checks), and I think IE9 will be able to handle this case as well.

cedricmoullet commented:
Thanks

2016-06-10 13:22 GMT+02:00 Gilbert Jeiziner <notifications@github.com>:

> I looked deeper in their code (which is written with the necessary
> checks), and I think IE9 will be able to handle this case as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/3199#issuecomment-225157643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AATgvPguqa9ibG-ZafQmt30WX9U8mWd9ks5qKUjegaJpZM4HuliU>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://twitter.com/cedricmoullet>
